test/clj/rems/test/browser.clj

```diff
@@ -0,0 +1,140 @@
+            [rems.standalone]))
+
+(def ^:dynamic *driver*
+  "Current driver")
+
+(def reporting-dir (doto (io/file "browsertest-errors")
+                     (.mkdirs)))
+
+(defn fixture-driver
+  "Executes a test running a driver.
+   Bounds a driver with the global *driver* variable."
+  [f]
+  (with-chrome-headless {} driver
+    (binding [*driver* driver]
+      (f))))
+
+(defn fixture-standalone [f]
+  (mount/start)
+  (migrations/migrate ["reset"] (select-keys rems.config/env [:database-url]))
+  (test-data/create-test-data!)
+  (f)
+  (mount/stop))
+
+(use-fixtures
+  :each ;; start and stop driver for each test
+  fixture-driver)
+
+(use-fixtures
+  :once
+  fixture-standalone)
+
+;;; basic navigation
+
+(defn login-as [username]
+  (doto *driver*
+    (set-window-size 1400 2000) ;; big enough to show the whole page without scrolling
+    (go "http://localhost:3001")
+    (screenshot (io/file reporting-dir "landing-page.png"))
+    (click-visible {:class "login-btn"})
+    (screenshot (io/file reporting-dir "login-page.png"))
+    (click-visible [{:class "users"} {:tag :a, :fn/text username}])
+    (wait-visible :logout)))
+
+(defn click-navigation-menu [link-text]
+  (click-visible *driver* [:big-navbar {:tag :a, :fn/text link-text}]))
+
+(defn go-to-catalogue []
+  (click-navigation-menu "Catalogue")
+  (wait-visible *driver* {:tag :h2, :fn/text "Catalogue"})
+  (wait-invisible *driver* {:css ".fa-spinner"}))
+
+(defn go-to-applications []
+  (click-navigation-menu "Applications")
+  (wait-visible *driver* {:tag :h2, :fn/text "Applications"})
+  (wait-invisible *driver* {:css ".fa-spinner"}))
+
+;;; catalogue page
+
+(defn add-to-cart [resource-name]
+  (click-visible *driver* [{:css "table.catalogue"}
+                           {:fn/text resource-name}
+                           {:xpath "./ancestor::tr"}
+                           {:css "button.add-to-cart"}]))
+
+(defn apply-for-resource [resource-name]
+  (click-visible *driver* [{:css "table.cart"}
+                           {:fn/text resource-name}
+                           {:xpath "./ancestor::tr"}
+                           {:css "button.apply-for-resource"}])
+  (wait-visible *driver* {:tag :h2, :fn/text "Application"}))
+
+;;; application page
+
+(defn fill-form-field [label text]
+  (let [id (get-element-attr *driver* [:form
+                                       {:tag :label, :fn/text label}]
+                             :for)]
+    ;; XXX: need to use `fill-human`, because `fill` is so quick that the form drops characters here and there
+    (fill-human *driver* {:id id} text)))
+
+(defn accept-license [label]
+  ;; XXX: assumes that the checkbox is unchecked
+  (click-visible *driver* [:licenses
+                           {:tag :a, :fn/text label}
+                           {:xpath "./ancestor::div[@class='license']"}
+                           {:css "input[type='checkbox']"}]))
+
+(defn send-application []
+  (click-visible *driver* :submit)
```

Comment:
Perhaps we could use a named class here?

Reply:
"submit" is the element id. But it would be better be "submit-application".

Reply:
Though that would conflict quite heavily with your branch, so I'll leave it for now.